lib/tpm2_session.c

```diff
@@ -164,3 +172,150 @@ tpm2_session *tpm2_session_new(TSS2_SYS_CONTEXT *sapi_context,
 
     return session;
 }
+
+#define SESSION_VERSION 1
+
+/*
+ * Checks that two types are equal in size.
+ *
+ * It works by leveraging the fact that C does not allow negative array sizes.
+ * If the sizes are equal, the boolean equality operator will return 1, thus
+ * a subtraction of 1 yields 0, which is a legal array size in C. In the false
+ * case (ie sizes not equal), 0 - 1 is -1, which will cause the compiler to
+ * complain.
+ */
+#define COMPILE_ASSERT_SIZE(a, b) \
+    typedef char WRONG_SIZE_##a[(sizeof(a) == sizeof(b)) - 1]
+
+// We check that the TSS library does not change sizes unbeknownst to us.
+COMPILE_ASSERT_SIZE(TPM2_HANDLE, UINT32);
+COMPILE_ASSERT_SIZE(TPMI_ALG_HASH, UINT16);
+COMPILE_ASSERT_SIZE(TPM2_SE, UINT8);
+
+tpm2_session *tpm2_session_restore(const char *path) {
+
+    tpm2_session *s = NULL;
+
+    FILE *f = fopen(path, "rb");
+    if (!f) {
+        LOG_ERR("Could not open path \"%s\", due to error: \"%s\"",
+                path, strerror(errno));
+        goto out;
+    }
+
+    uint32_t version;
+    bool result = files_read_header(f, &version);
+
+    TPM2_SE type;
+    result = files_read_bytes(f, &type, sizeof(type));
+    if (!result) {
+        LOG_ERR("Could not read session type");
+        goto out;
+    }
+
+    TPMI_ALG_HASH auth_hash;
+    result = files_read_16(f, &auth_hash);
+    if (!result) {
+        LOG_ERR("Could not read session digest algorithm");
+        goto out;
+    }
+
+    TPM2_HANDLE handle;
+    result = files_read_32(f, &handle);
+    if (!result) {
+        LOG_ERR("Could not read session handle");
+        goto out;
+    }
+
+    tpm2_session_data *d = tpm2_session_data_new(type);
+    if (!d) {
+        LOG_ERR("oom");
+        goto out;
+    }
+
+    tpm2_session_set_authhash(d, auth_hash);
+
+    s = tpm2_session_new(NULL, d);
```

Comment:
You have to check if tpm2_session_new() didn't return NULL, and free tpm2_session_data *d if that's the case.

Reply:
yes